### ISSUE TITLE
SSAFY

### ISSUE THREAD

**Issue by jyoungyun** [jyoungyun/Test#40]

**Comment by jyoungyun** [comment]
Q. Compiler, Backend Extension, Toolchain의 상호작용과 계층관계가 어떻게 되나요?
A.

1)
Toolchain 은 Backend-specific tool 까지 포함하고 있는 것이 맞으나, 이 프로젝트에서는 Backend 가 구체화되지 않았기 때문에 Toolchain 에 one-compiler 만 포함하는 것으로 하면 좋을 것 같습니다.
(만약, 시간적 여유가 된다면 dummy backend 를 구현하여 포함할 수도 있겠습니다.)

2)
Toolchain package 는 one-vscode 와 backend extension 의 도움 없이도 artifactory (debian source repository)를 통해 설치가 가능합니다.

3)
이 프로젝트는 ONE compiler 의 결과물을 활용하는 것으로써

**Comment by jyoungyun** [comment]
## Optimizer

- 목적
   - loss function 의 최소값을 찾는데, 이것을 효율적으로 찾기 위함

### Gradient Descent 의 원리

- 경사 하강법
- 원리
   - convex function(볼록함수) 라면
   - 기울기를 이용해 기울기의 반대 방향으로 따라 내려가다보면 언젠가는 기울기가 0 인 지점을 찾을 수 있다
- formula
   - ![gd_sgd_formula](https://media.github.sec.samsung.net/user/479/files/58e1aa0b-e360-49e0-8b63-30c489a499f8)
   - ```math
      w(t+1) = w(t) - \eta\frac{\partial}{\partial w}Cost(w)
      ```
   - $\eta : learning rate$
- 설명
   - loss function 을 함수로 나타낸다면 우리는 loss 가 가장 작은 최소값을 찾는 것이 목적
   - 값이 가장 작은 최소값을 찾기 위해서는 기울기가 0이 되는 방향으로 가다보면 찾을 수 있을것

### BGD (Batch Gradient Descent)

- full batch 경사 하강법
- 원리
   - 현재 위치에서 모든 자료(full batch)의 기울기를 계산해서 gradient descent 가 가장 큰 방향의 반대로 이동
- 설명
   - 여기서 입력 데이터 x 에 대한 편미분이 아닌 가중치 w 에 대한 편미분이라는 것을 기억
   - 따라서 모든 데이터에 대한 에러를 구한 뒤 이때의 가중치 w 에 대해 기울기를 계산하여 사용
- formula
   - 경사하강법과 동일
- 장점
   - 전체 데이터에 대해 업데이트가 한번에 실행되기 때문에 나중에 나오는 SGD 보다 업데이트 횟수가 적음
   - 전체적인 계산 횟수는 SGD 보다 적음
   - 전체 데이터에 대해 error gradient 를 계산하므로 optimal 로의 수렴이 안정적임
   - 병렬처리에 유리
- 단점
   - 한 스텝에 모든 학습 데이터 셋을 사용하므로 학습이 오래 걸림
   - 전체 학습 데이터에 대한 error 를 모델의 업데이트가 이루어지기 전까지 저장해야 하므로 더 많은 메모리 사용
   - local optimal 에 빠질 수 있음

### SGD (Stochastic Gradient Descent)

- 확률적 경사하강법
- 원리
   - 현재 위치에서 랜덤하게 추출한 하나의 데이터에 대해 기울기를 계산해서 gradient descent 가 가장 큰 방향의 반대로 이동
- 설명
   - 랜덤하게 추출된 하나의 데이터에 대해서 error gradient 를 계산
   - 수식적으로 BGD 와 동일, 다른 부분은 전체 데이터가 아닌 랜덤하게 추출된 하나의 데이터라는 점
- formula
   - 경사하강법과 동일
- 장점
   - 데이터 하나하나에 대해 민감하게 반응하므로 shooting 이 일어날 수 있어 local optimal 에 빠질 리스크가 적음
   - 한 step 에 걸리는 시간이 짧기 때문에 수렴 속도가 BGD 에 비해 빠름
- 단점
   - Global optimal 을 찾지 못할 가능성이 있음
   - 데이터를 한개씩 처리하므로 GPU 성능을 전부 활용할 수 없음

### MSGD (Mini-batch Gradient Descent)

- 미니 배치 경사 하강법
- 원리
   - 전체 데이터셋에서 뽑은 mini-batch 안의 데이터 m 개에 대해 기울기 m개를 계산해 그것의 평균 기울기를 통해 gradient descent 가 가장 큰 방향의 반대로 이동
- 설명
   - SGD 와 MSGD 는 서로 다른 알고리즘이지만 많이들 MSGD 를 SGD 로 사용하고 있음
   - 전체 데이터 셋을 여러개의 mini-batch로 나누어 한개의 mini-batch 마다 기울기를 구하고 모델 업데이트
   - batch_size 는 hyper-parameter 
      - 보통 2의 제곱수를 이용하는데, 많은 벡터 계산이 2의 제곱수가 입력될 때 빠르기 때문 (?)
- formula
   - 경사하강법과 동일
- 장점
   - BGD 보다 local optimal 에 빠질 리스크가 적음
   - SGD 보다 병렬처리에 유리
   - 전체 학습 데이터가 아닌 일부 학습 데이터만을 사용하므로 메모리 사용이 BGD 보다 적음
- 단점
   - batch-size 만큼 에러에 대한 내용을 저장해야 하므로 SGD 보다는 많음

### Momentum

- 원리
   - 현재의 data 로만 학습하는 것이 아니라 이전의 학습 결과도 반영하여 이동
- 설명
   - GD 의 경우 local optimal 에 빠지면 나올 수가 없음
   - 관성을 이용하면 local optimal 에 빠진 경우에도 가던 방향으로 가던 습성으로 인해 빠져나올 수 있는 여지를 줌
   - learning rate 외에 momentum 이라는 hyper-parameter 가 추가됨
   - momentum 값은 보통 0.9 에서 시작하며 0.95, 0.99 로 증가시키며 사용한다고 함
   - 통계학에서는 momentum 식이 ema(Exponential Moving Average) 혹은 EWMA(Exponential Weighted Moving Average)로 표현되기도 함 (지수평균의 의미)
      - $v_{0} = 0$ 으로 두고 $\sum_{}^{}$ 으로 묶으면
         - ![momentum_ema_formula](https://media.github.sec.samsung.net/user/479/files/fd491f4e-2cdc-475e-9f26-376764cdc043)
         - $v_{t} = \sum_{i=0}^{t} \alpha^{t-i}\nabla L(w_{t})$
- formula
   - ![momentum_formula](https://media.github.sec.samsung.net/user/479/files/2d7e0719-bcee-4f64-b1cd-f8bcb5d53b26)
   - ```math
      \begin{multiline}
      v_{t+1} = \alpha v_{t} - \eta\nabla L(w_{t}) \\ \\
      w_{t+1} = w_{t}+v_{t+1}
      \end{multiline}
      ```
   - $\alpha : momentum hyper-parameter$
- 장점
   - local optimal 문제에 강인함

### NAG (Nesterov Accelrated Gradient, Nestrerov Momentum)

- 원리
   - 이동할 방향을 미리 예측하고 그 방향으로 이동한 후에 gradient 를 계산함으로써 한단계를 미리 예측해 불필요한 이동을 줄임
   - ![nag_1](https://media.github.sec.samsung.net/user/479/files/6ed62d61-8da9-4666-9a53-4ac02f877330)
- 설명
   - 그다지 유명한 것은 아니지만 그래도 설명을 해보자면
   - momentum 이슈를 좀더 개선해서 Nesterov 라는 학자가 고안해 낸 알고리즘
   - momentum 값이 적용된 지점에서 gradient 값을 계산
   - 기존
      - 현재 위치에서 gradient 를 구한다
      - 이전 velocity 를 가져온다
      - 이 두가지 방향을 더한 값으로 step 을 진행
   - 이 방법
      - 이전 veolcity 를 가져온다
      - 이전 velocity 방향으로만 갔을 때의 지점에서 gradient 를 구한다
      - 이 두가지를 더한 방향으로 step 진행
   - 수식이 완전히 이해가 되지는 않음...
 
### AdaGrad

- 원리
   - 진행 횟수에 따라 learning rate 를 조절하는 옵션(learning rate decay)이 추가된 Gradient Descent 방법
      - step 별, 변수 별 각기 다른 learning rate 를 가짐
- 설명
   - GD 계열 방식은 모두 learning rate 가 고정이라 기울기를 반영하는 비율이 동일함
      - step 마다 상수, 변수마다 상수
   - local minimum 에 가까워질 수록 learning rate 를 줄여나가는 것이 더 효율적일수도 있으므로 learning rate 를 가변적으로 적용하기 위한 아이디어 
   - 많이 변화하지 않은 변수들은 learning rate 를 크게하고, 많이 변환한 변수들은 learning rate 를 적게 변화함
   - 많이 변화한 변수는 이미 최적값에 근접했을 것이라는 가정
   - 또한 모든 weight 에 대해 하나의 gradient 를 가지는 것이 아니고 각 weight 별로 gradient 를 가짐
- formula
   - ![adagrad_formula](https://media.github.sec.samsung.net/user/479/files/41116bca-cb56-4bad-839a-d40f54d76422)
   - ```math
      \begin{multiline}
      h_{t+1} = h_{t} + \nabla L(w_{t})\cdot\nabla L(w_{t}) \\ \\
      w_{t+1} = w_{t} - \eta \frac{1}{\sqrt{h_{t+1}}} \nabla L(w_{t})
      \end{multiline}
      ```
- 장점
   - SGD 에 비해 convex function 의 경우 global optimal 을 잘 찾을수있음
- 단점
   - 학습이 무한히 진행되면 어느순간 h 가 너무 커져서 더이상 학습이 안되는 문제가 있음
      - 이걸 해결한게 RMSProp
   - convex function 에서는 효과적이지만 non-convex function 의 경우에는 saddle point 에 갇힐 수 있음

### RMSProp

- 등장
   - AdaGrad 의 학습이 많이 진행될 경우 h 가 너무 커져서 더이상 학습이 안되는 단점을 보완한 방법
- 원리
   - 가중치 기울기를 단순히 누적시키는 것이 아니라 최신 기울기들이 더 반영되도록 하는 방법
- 설명
   - 새로운 가중치와 예전 가중치의 비율을 조절해서 최신 기울기를 더 많이 반영되도록 함
   - 즉, 그동안 가중치의 큰 변동사항이 있었어도 이번 batch로 인한 기울기가 완만하면 학습률을 높임
   - hyper paramter $\rho$ 를 추가하여 h 가 무한히 커지지 않게 함
- formula
   - ![rmsprop_formula](https://media.github.sec.samsung.net/user/479/files/a8b26d84-dbe3-4b27-a0c3-bebcc3c45aff)
    - ```math
      \begin{multiline}
      h_{t+1} = \rho h_{t} + (1 - \rho)\nabla L(w_{t})\cdot\nabla L(w_{t}) \\ \\
      w_{t+1} = w_{t} - \eta \frac{1}{\sqrt{h_{t+1}}} \nabla L(w_{t})
      \end{multiline}
      ```
- 장점
   - AdaGrad 의 학습이 안되는 문제를 해결
- 단점
   - momentum 처럼 global optimal 이 아닌 local optimal 에 빠질수 있음

### Adam (Adaptive Moment Estimation)

- 등장
   - 기존 RMSProp 과 momentum 는 처음 v 나 h 의 값이 0 이 되어  학습 초반에 0 으로 biased 되는 문제를 해결하기 위해 고안된 방법
- 원리
   - Momentun 과 RMSProp 를 합친 경사 하강법
   - RMSProp 의 특징인 gradient 의 제곱을 지수 평균한 값을 사용하며 momentum의 특징으로 gradient 를 제곱하지 않은 값을 사용하여 지수 평균을 구하고 수식에 활용
- 설명
   - 이 방식을 제안한 논문에 따르면 AdaGrad + RMSProp 를 combine 했다고 함
   - 하지만 RMSProp 은 AdaGrad 의 발전된 버전인걸?
   - 좀더 많은 문헌을 찾아보면 RMSProp + Momentum 이라는 말들도 많음
      - momentum 은 이전 속도에 관성을 주는 hyper-paramter 를 줘서 그걸 반영시키는 거였고
      - RMSProp 은 AdaGrad 의 많이 변동했을 경우 적게 움직이고 적게 움직였을 경우 많이 움직이는 것에서 각 비율이 동일하게 반영되는 것이 아니라 $\rho$ 와 $1-\rho$ 를 현재와 과거에 서로 다른 비율을 줘서 현재를 더 많이 반영시키도록 하는 방법
   - 기존 RMSProp 과 momentum 은 처음 v 나 h 값이 0 으로 설정되어 학습 초반에 0 으로 bias 되는 문제가 있음
      - ![momentum_rmsprop_formula](https://media.github.sec.samsung.net/user/479/files/55227d97-1f85-40ba-a0ec-38643dd2e867)
   - 1차 모멘텀인 m(momentum 의 v) 과 2차 모멘텀인 v(RMSProp 의 h)를 이용해 최적화를 진행, g 는 현재의 gradient descent 값
   - $m_{0} = 0$ 이어도 $m_{1} $leftarrow (1 - \beta_{1})g_{1} = 0.1g_{1} (\because m_{0} = 0)$
   - 현재로써는 이 방법이 제일로 널리 쓰이고 효과적이라고 함.
- formula
   -

**Comment by jyoungyun** [comment]
<details>

```json
{
  // See https://go.microsoft.com/fwlink/?LinkId=733558
  // for the documentation about the tasks.json format
  "version": "2.0.0",
  "tasks": [
    {
      "label": "ONE-Compiler: Build ONE Compiler",
      "type": "shell",
      "command": "DESTDIR=$(pwd)/overlay ./nncc build -- install"
    },
    {
      "label": "ONE-Compiler: Configure ONE Compiler",
      "type": "shell",
      "command": "./nncc configure"
    },
    {
      "label": "ONE-Runtime: Host X86 Docker Build (Debug)",
      "type": "shell",
      "command": "./nnas docker-run make -f Makefile.template"
    },
    {
      "label": "ONE-Runtime: Host X86 Docker Build (Release)",
      "type": "shell",
      "command": "./nnas docker-run make -f Makefile.template BUILD_TYPE=release"
    },
    {
      "label": "ONE-Runtime: ARM Cross Build (Debug)",
      "type": "shell",
      "command": "ROOTFS_DIR=$(pwd)/tools/cross/rootfs/arm/ CROSS_BUILD=1 TARGET_ARCH=armv7l make -f Makefile.template"
    },
    {
      "label": "ONE-Runtime: ARM Cross Build (Release)",
      "type": "shell",
      "command": "ROOTFS_DIR=$(pwd)/tools/cross/rootfs/arm/ CROSS_BUILD=1 TARGET_ARCH=armv7l make -f Makefile.template BUILD_TYPE=release"
    },
    {
      "label": "ONE-Runtime: ARM64 Cross Build (Debug)",
      "type": "shell",
      "command": "ROOTFS_DIR=$(pwd)/tools/cross/rootfs/aarch64/ CROSS_BUILD=1 TARGET_ARCH=aarch64 make -f Makefile.template"
    },
    {
      "label": "ONE-Runtime: ARM64 Cross Build (Release)",
      "type": "shell",
      "command": "ROOTFS_DIR=$(pwd)/tools/cross/rootfs/aarch64/ CROSS_BUILD=1 TARGET_ARCH=aarch64 make -f Makefile.template BUILD_TYPE=release"
    },
    // {
    //   "label": "ONE-Runtime : Host X86(w/Trix) Build (Debug)",
    //   "type": "shell",
    //   "command": "OPTIONS='-DTRIX_ENGINE_PREFIX=/opt/trinity -DBUILD_NPUD=ON -DENVVAR_NPUD_CONFIG=ON' make -f Makefile.template all install | tee ./Product/ONE-Runtime_Host_X86_Trix_Build_Debug.out"
    // },
    // {
    //   "label": "ONE-Runtime : Host X86(w/Trix) Build (Release)",
    //   "type": "shell",
    //   "command": "BUILD_TYPE=release OPTIONS='-DTRIX_ENGINE_PREFIX=/opt/trinity -DBUILD_NPUD=ON -DENVVAR_NPUD_CONFIG=ON' make -f Makefile.template all install | tee ./Product/ONE-Runtime_HOST_X86_Trix_Build_Release.out"
    // },
    // {
    //   "label": "ONE-Runtime: ARM Docker Build (Debug)",
    //   "type": "shell",
    //   "command": "DOCKER_IMAGE_NAME=docker-remote.bart.sec.samsung.net/nnfw/one-devtools:focal ./nnas docker-run make -f Makefile.template CROSS_BUILD=1 TARGET_ARCH=armv7l ROOTFS_DIR=${PWD}/tools/cross/rootfs/arm install"
    // },
    // {
    //   "label": "ONE-Runtime: ARM Docker Build (Debug)",
    //   "type": "shell",
    //   "command": "CROSS_BUILD=1 TARGET_ARCH=armv7l ROOTFS_DIR=${PWD}/tools/cross/rootfs/arm DOCKER_IMAGE_NAME=docker-remote.bart.sec.samsung.net/nnfw/one-devtools:focal ./nnas docker-run make -f Makefile.template install"
    // },
    // {
    //   "label": "ONE-Runtime: AARCH64 Docker Build (Debug) (Deprecated)",
    //   "type": "shell",
    //   "command": "ROOTFS_DIR=${PWD}/tools/cross/rootfs/aarch64 DOCKER_IMAGE_NAME=nnfw/one-devtools:bionic ./infra/scripts/docker_build_cross_aarch64_runtime.sh"
    // },
    // {
    //   "label": "ONE-Runtime : X86(w/Trix) Docker Build (Release) (Deprecated)",
    //   "type": "shell",
    //   "command": "DOCKER_IMAGE_NAME=rs7-runtimentools-docker-local.bart.sec.samsung.net/one-devtools:bionic-trix-engine BUILD_OPTIONS=' -DTRIX_ENGINE_PREFIX=/opt/trinity -DBUILD_NPUD=ON -DENVVAR_NPUD_CONFIG=ON' ./infra/scripts/docker_build_test_x64.sh"
    // },
    // {
    //   "label": "ONE-Runtime: X86 Docker Build & Test (Debug) (Deprecated)",
    //   "type": "shell",
    //   "command": "DOCKER_IMAGE_NAME=nnfw/one-devtools:bionic ./infra/scripts/docker_build_test_x64.sh"
    // },
    // {
    //   "label": "ONE-Runtime: Android Docker Build (Debug) (Deprecated)",
    //   "type": "shell",
    //   "command": "DOCKER_IMAGE_NAME=nnfw/one-devtools:bionic ./infra/scripts/docker_build_cross_android.sh"
    // },

    // {
    //   "label": "ONE-Runtime: Tizen Docker Build (Release) (Deprecated)",
    //   "type": "shell",
    //   "command": "ROOTFS_DIR=${PWD}/tools/cross/rootfs/armel DOCKER_IMAGE_NAME=docker-remote.bart.sec.samsung.net/nnfw/tizen-gbs:latest ./infra/scripts/docker_build_tizen_gbs.sh"
    // }
  ]
}
```

</details>
